File: backend/app/api/v1/routes/claims.py
from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import CurrentUserDep, get_db
from app.schemas.claims import (
    ClaimCompareData,
    ClaimSummaryData,
    CompletenessScoreData,
    ReadyToFileData,
)
from app.schemas.common import ApiResponse
from app.services.borang_be import BorangBeService
from app.services.claims import ClaimsService
from app.services.engagement import EngagementService
from app.services.export import ExportService

router = APIRouter(prefix="/claims", tags=["claims"])


def _claims_service(db: AsyncSession = Depends(get_db)) -> ClaimsService:
    return ClaimsService(db)


def _borang_be_service(db: AsyncSession = Depends(get_db)) -> BorangBeService:
    return BorangBeService(db)


def _export_service(db: AsyncSession = Depends(get_db)) -> ExportService:
    return ExportService(db)


def _engagement_service(db: AsyncSession = Depends(get_db)) -> EngagementService:
    return EngagementService(db)


@router.get("/summary", response_model=ApiResponse[ClaimSummaryData])
async def get_claim_summary(
    current_user: CurrentUserDep,
    tax_year: int | None = Query(default=None, ge=2000, le=2100),
    service: ClaimsService = Depends(_claims_service),
) -> ApiResponse[ClaimSummaryData]:
    data = await service.get_summary(current_user, tax_year=tax_year)
    return ApiResponse(success=True, data=data, message=None)


@router.get("/compare", response_model=ApiResponse[ClaimCompareData])
async def get_claim_comparison(
    current_user: CurrentUserDep,
    tax_year: int | None = Query(default=None, ge=2000, le=2100),
    service: ClaimsService = Depends(_claims_service),
) -> ApiResponse[ClaimCompareData]:
    data = await service.get_comparison(current_user, tax_year=tax_year)
    return ApiResponse(success=True, data=data, message=None)


@router.get("/ready-to-file", response_model=ApiResponse[ReadyToFileData])
async def get_ready_to_file(
    current_user: CurrentUserDep,
    tax_year: int | None = Query(default=None, ge=2000, le=2100),
    service: BorangBeService = Depends(_borang_be_service),
) -> ApiResponse[ReadyToFileData]:
    data = await service.get_ready_to_file(current_user, tax_year=tax_year)
    return ApiResponse(success=True, data=data, message=None)


@router.get("/completeness", response_model=ApiResponse[CompletenessScoreData])
async def get_completeness_score(
    current_user: CurrentUserDep,
    tax_year: int | None = Query(default=None, ge=2000, le=2100),
    service: EngagementService = Depends(_engagement_service),
) -> ApiResponse[CompletenessScoreData]:
    data = await service.get_completeness_score(current_user, tax_year=tax_year)
    return ApiResponse(success=True, data=data, message=None)


@router.get("/export-zip")
async def export_receipts_zip(
    current_user: CurrentUserDep,
    tax_year: int = Query(ge=2000, le=2100),
    service: ExportService = Depends(_export_service),
) -> Response:
    content, filename = await service.build_receipts_zip(
        current_user,
        tax_year=tax_year,
    )
    return Response(
        content=content,
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: frontend/src/lib/api/claims.ts
import "server-only";

import { apiFetch } from "@/lib/api/client";
import { requireSessionCookieHeader } from "@/lib/api/session";
import type {
  ClaimCompareData,
  ClaimSummaryData,
  CompletenessScoreData,
  ReadyToFileData,
} from "@/lib/api/types";

export async function fetchClaimSummary(taxYear?: number) {
  const cookie = await requireSessionCookieHeader();
  const query = taxYear ? `?tax_year=${taxYear}` : "";

  const result = await apiFetch<ClaimSummaryData>(`/api/v1/claims/summary${query}`, {
    cookie,
  });

  return result;
}

export async function fetchClaimComparison(taxYear?: number) {
  const cookie = await requireSessionCookieHeader();
  const query = taxYear ? `?tax_year=${taxYear}` : "";

  const result = await apiFetch<ClaimCompareData>(`/api/v1/claims/compare${query}`, {
    cookie,
  });

  return result;
}

export async function fetchReadyToFile(taxYear?: number) {
  const cookie = await requireSessionCookieHeader();
  const query = taxYear ? `?tax_year=${taxYear}` : "";

  return apiFetch<ReadyToFileData>(`/api/v1/claims/ready-to-file${query}`, {
    cookie,
  });
}

export async function fetchCompletenessScore(taxYear?: number) {
  const cookie = await requireSessionCookieHeader();
  const query = taxYear ? `?tax_year=${taxYear}` : "";

  return apiFetch<CompletenessScoreData>(`/api/v1/claims/completeness${query}`, {
    cookie,
  });
}

export const getClaimSummary = fetchClaimSummary;
export const getYearComparison = fetchClaimComparison;
export const getReadyToFile = fetchReadyToFile;
export const getCompletenessScore = fetchCompletenessScore;

export { exportZip } from "@/lib/types/claims";
export { getExportZipUrl } from "@/lib/api/export-urls";
